File: dashboard/test/ui/features/step_definitions/applab.rb
# Which lesson of allthethings.script contains the App Lab levels; this way we
# only have to update in one place if this changes.
APPLAB_ALLTHETHINGS_LESSON = 18

When /^I add code for a canvas and a button$/ do
  code =
    "createCanvas('my_canvas', 320, 480);\\n" \
    "button('my_button', 'ButtonText');"
  add_code_to_editor(code)
end

And /^Applab HTML has a button$/ do
  code = @browser.execute_script "return Applab.levelHtml"
  expect(/button/.match(code).nil?).to be(false)
end

And /^Applab HTML has no button$/ do
  code = @browser.execute_script "return Applab.levelHtml"
  expect(/button/.match(code).nil?).to be(true)
end

Given /^I start a new Applab project/ do
  steps <<-STEPS
    And I am on "http://studio.code.org/projects/applab/new"
    And I rotate to landscape
    And I wait for the page to fully load
    And element "#runButton" is visible
    And element "#codeModeButton" is visible
    And element "#designModeButton" is visible
    And element "#dataModeButton" is visible
  STEPS
end

Given /^I am on the (\d+)(?:st|nd|rd|th)? App ?Lab test level$/ do |level_index|
  steps <<-STEPS
    And I am on "http://studio.code.org/s/allthethings/lessons/#{APPLAB_ALLTHETHINGS_LESSON}/levels/#{level_index}"
    And I rotate to landscape
    And I wait for the page to fully load
  STEPS
end

When /^I switch to design mode$/ do
  steps <<-STEPS
    When I press "designModeButton"
    And I wait to see Applab design mode
  STEPS
end

When /^I switch to data mode$/ do
  steps <<-STEPS
    When I press "dataModeButton"
    And I wait to see Applab data mode
  STEPS
end

When /^I switch to code mode$/ do
  steps <<-STEPS
    When I press "codeModeButton"
    And I wait to see Applab code mode
  STEPS
end

And /^I wait to see Applab design mode$/ do
  wait = Selenium::WebDriver::Wait.new(timeout: 10)
  wait.until {@browser.execute_script("return $('#designWorkspace').css('display') == 'block';")}
end

And /^I wait to see Applab data mode$/ do
  wait = Selenium::WebDriver::Wait.new(timeout: 10)
  wait.until {@browser.execute_script("return $('#dataWorkspaceWrapper').css('display') == 'block';")}
end

And /^I wait to see Applab code mode$/ do
  wait = Selenium::WebDriver::Wait.new(timeout: 30)
  wait.until {@browser.execute_script("return $('#codeWorkspaceWrapper').css('display') == 'block';")}
end

# Step for dragging an Applab design mode element into the applab visualization.
# Use with element type strings from ElementType (library.js)
When /^I drag a (\w+) into the app$/ do |element_type|
  drag_script = %Q{
    var element = $("[data-element-type='#{element_type}']");
    var screenOffset = element.offset();
    var mousedown = $.Event("mousedown", {
      which: 1,
      pageX: screenOffset.left,
      pageY: screenOffset.top
    });
    var drag = $.Event("mousemove", {
      pageX: $("#visualization").offset().left + 15,
      pageY: $("#visualization").offset().top
    });
    var mouseup = $.Event('mouseup', {
      pageX: $("#visualization").offset().left + 15,
      pageY: $("#visualization").offset().top
    });
    element.trigger(mousedown);
    $(document).trigger(drag);
    $(document).trigger(mouseup);
  }
  @browser.execute_script(drag_script)
end

When /^I navigate to the embedded version of my project$/ do
  steps <<-STEPS
    When I open the share dialog
    And I click selector "#project-share a:contains('Show advanced options')"
    And I click selector "#project-share li:contains('Embed')"
    And I copy the embed code into a new document
  STEPS
end

When /^I navigate to the embedded version of my project with source hidden$/ do
  steps <<-STEPS
    When I open the share dialog
    And I click selector "#project-share a:contains('Show advanced options')"
    And I click selector "#project-share li:contains('Embed')"
    And I click selector "#project-share label:contains('Hide ability to view code')"
    And I copy the embed code into a new document
  STEPS
end

Then(/^the palette has (\d+) blocks$/) do |num_blocks|
  expect(@browser.execute_script("return $('.droplet-palette-canvas > g').length")).to eq(num_blocks.to_i)
end

Then(/^the droplet code is "([^"]*)"$/) do |code|
  code.gsub!("\\n", "\n")
  expect(@browser.execute_script("return Applab.getCode()")).to eq(code)
end

And /^I append text to droplet "([^"]*)"$/ do |text|
  script = %Q{
    var aceEditor = window.__TestInterface.getDroplet().aceEditor;
    aceEditor.navigateFileEnd();
    aceEditor.textInput.focus();
    aceEditor.onTextInput("#{text}");
  }
  @browser.execute_script(script)
end

def set_nth_input(n, value)
  elements = @browser.find_elements(:css, '#design-properties input')
  # For some reason, the test machine seemed to stop responding to :delete. Even
  # stranger, on my localhost, if I do a bunch of backspaces without following
  # them with a delete, the press_keys(value) is ignored. By having both here,
  # things seem to work both on test and in development
  press_keys(elements[n], "\b") # backspace
  press_keys(elements[n], "\b") # backspace
  press_keys(elements[n], "\b") # backspace
  press_keys(elements[n], "\b") # backspace
  press_keys(elements[n], ":delete")
  press_keys(elements[n], ":delete")
  press_keys(elements[n], ":delete")
  press_keys(elements[n], ":delete")
  press_keys(elements[n], value)
end

And /^I set input "([^"]*)" to "([^"]*)"$/ do |type, value|
  case type
  when 'xpos'
    # first key press will just clear the current value
    set_nth_input(3, value)
  when 'ypos'
    set_nth_input(4, value)
  else
    raise 'Unknown type'
  end
end

And /^I set groupable input "([^"]*)" to "([^"]*)"$/ do |type, value|
  case type
  when 'xpos'
    # first key press will just clear the current value
    set_nth_input(4, value)
  when 'ypos'
    set_nth_input(5, value)
  else
    raise 'Unknown type'
  end
end

And /^I delete the current design mode element$/ do
  elements = @browser.find_elements(:css, '#design-properties button')
  elements[0].click
end

def drag_grippy(element_js, delta_x, delta_y)
  script = get_mouse_event_creator_script

  script += %Q{
    var element = #{element_js};
    var start = {
      x: element.offset().left,
      y: element.offset().top
    };
    var end = {
      x: start.x + #{delta_x},
      y: start.y + #{delta_y}
    }
    var mousedown = createMouseEvent('mousedown', start.x, start.y);
    var drag = createMouseEvent('mousemove', end.x, end.y);
    var mouseup = createMouseEvent('mouseup', end.x, end.y);

    element[0].dispatchEvent(mousedown);
    element[0].dispatchEvent(drag);
    element[0].dispatchEvent(mouseup);
  }
  # Run the script and then wait a little bit of time to give the UI a chance to reflow.
  @browser.execute_script(script, 0.5)
end

And /^I drag the instructions grippy by ([-|\d]+) pixels$/ do |delta|
  drag_grippy('$(".fa-ellipsis-h").eq(0).parent()', 0, delta)
end

And /^I drag the visualization grippy by ([-|\d]+) pixels$/ do |delta|
  drag_grippy('$("#visualizationResizeBar")', delta, 0)
end

And /^I hover over element with id "([^"]*)"$/ do |element_id|
  script = get_mouse_event_creator_script
  script += get_scale_script

  script += %Q{
    var element = $("#" + "#{element_id}");
    var scale = getScale(element[0]);
    var x = element.offset().left + (5 * scale);
    var y = element.offset().top + (5 * scale);

    var mousemove = createMouseEvent('mousemove', x, y);

    element[0].dispatchEvent(mousemove);
  }

  @browser.execute_script(script)
end

And /^I hover over the screen at xpos ([\d]+) and ypos ([\d]+)$/ do |xpos, ypos|
  script = get_mouse_event_creator_script
  script += get_scale_script

  script += %Q{
    var visualization = $("#visualizationOverlay");
    var transform = window.getComputedStyle(visualization[0]).transform;
    var scale = parseFloat(/matrix\\(([^,]*),/.exec(transform)[1]);
    var x = visualization.offset().left + (#{xpos} * scale);
    var y = visualization.offset().top + (#{ypos} * scale);
    var mousemove = createMouseEvent('mousemove', x, y);

    visualization[0].dispatchEvent(mousemove);
  }

  @browser.execute_script(script)
end

def get_mouse_event_creator_script
  return <<-FUNCTION
    function createMouseEvent(type, clientX, clientY) {
      var evt;
      var e = {
        bubbles: true,
        cancelable: (type != "mousemove"),
        view: window,
        detail: 0,
        screenX: undefined,
        screenY: undefined,
        clientX: clientX,
        clientY: clientY,
        ctrlKey: false,
        altKey: false,
        shiftKey: false,
        metaKey: false,
        button: 0,
        relatedTarget: undefined
      };
      if (typeof( document.createEvent ) == "function") {
        evt = document.createEvent("MouseEvents");
        evt.initMouseEvent(type,
          e.bubbles, e.cancelable, e.view, e.detail,
          e.screenX, e.screenY, e.clientX, e.clientY,
          e.ctrlKey, e.altKey, e.shiftKey, e.metaKey,
          e.button, document.body.parentNode);
      } else if (document.createEventObject) {
        evt = document.createEventObject();
        for (var prop in e) {
          evt[prop] = e[prop];
        }
        evt.button = { 0:1, 1:4, 2:2 }[evt.button] || evt.button;
      }
      return evt;
    };
  FUNCTION
end

def get_scale_script
  return <<-FUNCTION
    function getScale(element) {
      return element.getBoundingClientRect().width / element.offsetWidth;
    };
  FUNCTION
end

And /^I save the project$/ do
  script = <<-SCRIPT
    Applab.serializeAndSave();
  SCRIPT

  @browser.execute_script(script)
end

And /^I drag element "([^"]*)" ([\d]+) horizontally and ([\d]+) vertically$/ do |element_id, delta_x, delta_y|
  script = get_mouse_event_creator_script
  script += get_scale_script

  script += %Q{
    var element = $("#{element_id}");
    var scale = getScale(element[0]);

    var start = {
      x: element.offset().left,
      y: element.offset().top
    };

    var end = {
      x: start.x + (#{delta_x} * scale),
      y: start.y + (#{delta_y} * scale)
    }
    var mousedown = createMouseEvent('mousedown', start.x, start.y);
    var drag = createMouseEvent('mousemove', end.x, end.y);
    var mouseup = createMouseEvent('mouseup', end.x, end.y);

    element[0].dispatchEvent(mousedown);
    element[0].dispatchEvent(drag);
    element[0].dispatchEvent(mouseup);
  }

  @browser.execute_script(script)
end

And /^I open the debug console$/ do
  steps 'And I click selector "#debug-area-header .fa-chevron-circle-up" if it exists'
end

Then(/^the share link includes "([^"]*)"$/) do |expected_text|
  share_link_input = nil
  wait_short_until do
    share_link_input = @browser.find_element(:css, '#sharing-input')
  end
  expect(share_link_input.attribute('value')).to include(expected_text)
end

table_name = nil
Then /^I save the table name from element "([^"]*)"$/ do |selector|
  table_name = @browser.execute_script("return $('#{selector}').text()")
end

Then /^I wait until (?:element )?"([^"]*)" (?:has|contains) the saved table name$/ do |selector|
  wait_until {@browser.execute_script("return $(#{selector.dump}).text();").include? table_name}
end

File: frontend/packages/apps-e2e-tests/tests/applab/AppLab.ts
import {expect, type Locator, type Page} from '@playwright/test';

/**
 * Page Object for App Lab — the Droplet/JS-based lab at lesson 18 of
 * allthethingscourse, or free projects at /projects/applab.
 *
 * App Lab uses a three-mode toolbar (Code / Design / Data) and a JS runtime
 * that runs inside #divApplab.  This POM covers the toolbar, mode switching,
 * the debug console, and the data-mode workspace — enough for the core
 * smoke scenarios.
 */
export class AppLab {
  /** Underlying Playwright page. */
  readonly page: Page;

  /** Run button — starts the user's program. */
  readonly runButton: Locator;

  /** Design Mode toggle button. */
  readonly designModeButton: Locator;

  /** Data Mode toggle button. */
  readonly dataModeButton: Locator;

  /** App canvas — #divApplab — contains all running app elements. */
  readonly appCanvas: Locator;

  /** Design workspace — #designWorkspace — visible in design mode. */
  readonly designWorkspace: Locator;

  /** Data workspace wrapper — #dataWorkspaceWrapper — visible in data mode. */
  readonly dataWorkspace: Locator;

  /** #codeModeButton — switches toolbar back to code view from design/data mode. */
  readonly codeModeButton: Locator;

  /** #codeWorkspaceWrapper — visible when in code mode. */
  readonly codeWorkspaceWrapper: Locator;

  /** #show-code-header — blocks ↔ text (ACE) mode toggle inside the code editor. */
  readonly showCodeHeader: Locator;

  /** Key-value output label rendered by getKeyValue/setKeyValue blocks. */
  readonly keyValueLabel: Locator;

  /** Record output label rendered by createRecord/readRecord blocks. */
  readonly recordLabel: Locator;

  /** Data library container — #data-library-container — lists available tables. */
  readonly dataLibraryContainer: Locator;

  /** Data table grid — #dataTable — visible after selecting a table. */
  readonly dataTable: Locator;

  /** Debug console output — #debug-output — accumulates console.log lines. */
  readonly consoleOutput: Locator;

  /** Reset button — stops the running program and restores initial state. */
  readonly resetButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.runButton = page.locator('#runButton');
    this.resetButton = page.locator('#resetButton');
    this.designModeButton = page.locator('#designModeButton');
    this.dataModeButton = page.locator('#dataModeButton');
    this.codeModeButton = page.locator('#codeModeButton');
    this.codeWorkspaceWrapper = page.locator('#codeWorkspaceWrapper');
    this.showCodeHeader = page.locator('#show-code-header');
    this.appCanvas = page.locator('#divApplab');
    this.designWorkspace = page.locator('#designWorkspace');
    this.dataWorkspace = page.locator('#dataWorkspaceWrapper');
    this.keyValueLabel = page.locator('#keyValueLabel');
    this.recordLabel = page.locator('#divApplab #recordLabel');
    this.dataLibraryContainer = page.locator('#data-library-container');
    this.dataTable = page.locator('#dataTable');
    this.consoleOutput = page.locator('#debug-output');
  }

  /**
   * Wait until App Lab is interactive enough to interact with.
   * Only #runButton is guaranteed on all level types (course levels lack the
   * design/data mode buttons).  Project pages (/projects/applab) do render
   * all three toolbar buttons, but they appear after #runButton.
   */
  async waitForReady(): Promise<void> {
    await expect(this.runButton).toBeVisible({timeout: 60_000});
  }

  /**
   * Click Design Mode and wait for the design workspace to appear.
   * Mirrors "I switch to design mode" from applab.rb.
   */
  async switchToDesignMode(): Promise<void> {
    await this.designModeButton.click();
    await expect(this.designWorkspace).toBeVisible();
  }

  /**
   * Click Data Mode and wait for the data workspace wrapper to appear.
   * Mirrors "I switch to data mode" from applab.rb.
   */
  async switchToDataMode(): Promise<void> {
    await this.dataModeButton.click();
    await expect(this.dataWorkspace).toBeVisible();
  }

  /**
   * Open the debug console by clicking the expand chevron if it is visible.
   * Mirrors "I open the debug console" from applab.rb:
   *   I click selector "#debug-area-header .fa-circle-chevron-up" if it exists
   */
  async openDebugConsole(): Promise<void> {
    const chevron = this.page.locator(
      '#debug-area-header .fa-circle-chevron-up',
    );
    if (await chevron.isVisible()) {
      await chevron.click();
    }
  }

  /** Click the run button to execute the current program. */
  async run(): Promise<void> {
    await this.runButton.click();
  }

  /**
   * Wait for App Lab's visible save status after a save-triggering action.
   * This observes the same readiness a user sees: catch "Saving..." when the
   * transition is visible, then require "Saved".  If the save is too quick to
   * catch, a changed saved timestamp is equivalent readiness.
   *
   * @param action - user action that triggers an App Lab source save
   * @param options - set expectSave false when the action may be a no-op
   */
  async waitForUiSaveAfter(
    action: () => Promise<void>,
    {expectSave = true}: {expectSave?: boolean} = {},
  ): Promise<void> {
    const previousTimestamp = await this.currentSavedTimestamp();

    await action();

    const saveStatus = this.page.locator('.project_updated_at');
    const saveError = this.page.locator('.project-save-error');
    const savingIndicator = saveStatus.filter({hasText: 'Saving'});

    if (expectSave) {
      const transition = await Promise.race([
        savingIndicator
          .waitFor({state: 'visible', timeout: 60_000})
          .then(() => 'saving' as const),
        expect
          .poll(() => this.currentSavedTimestamp(), {timeout: 60_000})
          .not.toBe(previousTimestamp)
          .then(() => 'saved' as const),
        saveError
          .waitFor({state: 'visible', timeout: 60_000})
          .then(() => 'error' as const),
      ]);

      if (transition === 'error') {
        throw new Error('App Lab reported a visible save error');
      }
    } else {
      await savingIndicator
        .waitFor({state: 'visible', timeout: 5_000})
        .catch(() => undefined);
    }

    const result = await Promise.race([
      expect(saveStatus)
        .toContainText('Saved', {timeout: 60_000})
        .then(() => 'saved' as const),
      saveError
        .waitFor({state: 'visible', timeout: 60_000})
        .then(() => 'error' as const),
    ]);

    if (result === 'error') {
      throw new Error('App Lab reported a visible save error');
    }
  }

  /**
   * Return the dateTime value displayed by the App Lab save indicator.
   */
  private async currentSavedTimestamp(): Promise<string | null> {
    return this.page
      .locator('.project_updated_at time')
      .getAttribute('datetime', {timeout: 1_000})
      .catch(() => null);
  }

  /**
   * Wait for the key-value output label to become visible.
   * Level 18/8: getKeyValue/setKeyValue blocks print to #keyValueLabel on success.
   */
  async waitForKeyValueLabel(): Promise<void> {
    await expect(this.keyValueLabel).toBeVisible({timeout: 30_000});
  }

  /**
   * Wait for the data library container to become visible after entering data mode.
   * The library loads asynchronously; this blocks until the table list is ready.
   */
  async waitForDataLibrary(): Promise<void> {
    await expect(this.dataLibraryContainer).toBeVisible({timeout: 30_000});
  }

  /**
   * Enter Data mode and wait until a data-table link is visible.
   * App Lab can render Data mode before level-defined tables reach the Data
   * Browser.  Agent Browser verified that the empty Data Tables list does not
   * update in place; switching back to Code and re-entering Data exposes the
   * table once App Lab is ready.  The visible table link is the readiness
   * signal replacing Cucumber's fixed delay.
   *
   * @param name - exact table name text as shown in the Data Tables list
   */
  async switchToDataModeWithTable(name: string): Promise<void> {
    await expect(async () => {
      if (await this.dataWorkspace.isVisible()) {
        await this.switchToCodeMode();
      }

      await this.switchToDataMode();
      await this.waitForDataLibrary();
      await expect(
        this.page.locator('#dataTablesBody a', {hasText: name}),
      ).toBeVisible({
        timeout: 2_000,
      });
    }).toPass({
      intervals: [250, 500, 1_000],
      timeout: 30_000,
    });
  }

  /**
   * Click a table name link in the data library and wait for the data grid.
   * Call waitForDataLibrary() before this to ensure the library is loaded.
   *
   * @param name - exact table name text as shown in the library list
   */
  async selectDataTable(name: string): Promise<void> {
    await this.page.locator('a', {hasText: name}).click();
    await expect(this.dataTable).toBeVisible({timeout: 15_000});
  }

  /**
   * Assert that a cell with the given text is visible in the current data table.
   * Call selectDataTable() first to open the table.
   *
   * @param text - cell text to look for (exact match)
   */
  async expectDataTableCell(text: string): Promise<void> {
    await expect(this.page.locator('td', {hasText: text})).toBeVisible();
  }

  /**
   * Switch from design or data mode back to code mode.
   * Clicks #codeModeButton and waits for #codeWorkspaceWrapper to appear.
   */
  async switchToCodeMode(): Promise<void> {
    await this.codeModeButton.click();
    await expect(this.codeWorkspaceWrapper).toBeVisible();
  }

  /**
   * Ensure the Droplet editor is in text (ACE) mode.
   * Checks #show-code-header button text; if "Show Text" clicks it and waits
   * for the droplet gutter to disappear (text mode signal from droplet_steps.rb).
   */
  async ensureTextMode(): Promise<void> {
    await this.showCodeHeader.waitFor({state: 'visible', timeout: 15_000});
    const text = await this.showCodeHeader.textContent();
    if (text?.trim() === 'Show Text') {
      await this.showCodeHeader.click();
      await this.page.waitForFunction(
        () => {
          const el = document.querySelector(
            '.droplet-gutter > div',
          ) as HTMLElement | null;
          return !el || getComputedStyle(el).display === 'none';
        },
        {timeout: 10_000},
      );
    }
  }

  /**
   * Append code to the Droplet ACE editor. Must call ensureTextMode() first.
   * Uses __TestInterface.getDroplet().aceEditor.onTextInput(), matching the
   * Cucumber `I append text to droplet "..."` step from droplet_steps.rb.
   *
   * @param code - source text to append at the end of the editor
   */
  async appendCode(code: string): Promise<void> {
    await expect(async () => {
      if (!(await this.dropletContents()).includes(code)) {
        await this.page.evaluate((c: string) => {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const ace = (window as any).__TestInterface.getDroplet().aceEditor;
          ace.navigateFileEnd();
          ace.textInput.focus();
          ace.onTextInput(c);
        }, code);
      }

      await expect.poll(() => this.dropletContents()).toContain(code);
    }).toPass({
      intervals: [250, 500, 1_000],
      timeout: 15_000,
    });
  }

  /**
   * Insert code at the current ACE cursor position. Must call ensureTextMode() first.
   * Mirrors `I add code "..." to ace editor` from droplet_steps.rb, which does
   * NOT navigate to file end — it inserts at the current cursor position.  After
   * a page reload the cursor defaults to position 0, so this effectively prepends.
   *
   * @param code - source text to insert at the current cursor position
   */
  async insertCodeAtCursor(code: string): Promise<void> {
    await this.page.evaluate((c: string) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const ace = (window as any).__TestInterface.getDroplet().aceEditor;
      ace.textInput.focus();
      ace.onTextInput(c);
    }, code);
  }

  /** Return the current text-mode Droplet source. */
  private async dropletContents(): Promise<string> {
    return this.page.evaluate(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return (window as any).__TestInterface.getDroplet().aceEditor.getValue();
    });
  }

  /**
   * Open the project share dialog, read the share URL from
   * #sharing-dialog-copy-button (mirrors `I save the share URL` from
   * project_steps.rb), close the dialog, and return the URL as a relative
   * path suitable for page.goto().
   *
   * Forces a save first via dashboard.project.save() so the share page
   * reflects the current code/design — autosave debounce is not reliable
   * enough to guarantee this without the explicit flush.
   */
  async getShareUrlFromDialog(): Promise<string> {
    // Open the share dialog (mirrors "I open the project share dialog"
    // from project_steps.rb: click .project_share, wait for dialog).
    await this.page.locator('.project_share').first().click();
    await this.page
      .locator('#project-share')
      .waitFor({state: 'visible', timeout: 15_000});

    // Read the share URL from the copy button's value attribute.
    // #sharing-dialog-copy-button is a MuiButton (<button> element) with the
    // URL in its value= attribute.  inputValue() only works for input/select,
    // so use getAttribute() instead — mirrors Cucumber's .value JS access.
    const copyButton = this.page.locator('#sharing-dialog-copy-button');
    await copyButton.waitFor({state: 'visible', timeout: 10_000});
    const fullUrl = await copyButton.getAttribute('value');
    if (!fullUrl) {
      throw new Error('share dialog copy button did not expose a URL value');
    }

    // Close dialog.
    await this.page.keyboard.press('Escape');
    await this.page
      .locator('#project-share')
      .waitFor({state: 'hidden', timeout: 10_000});

    // Strip origin so the path works against the test baseURL.
    return fullUrl.replace(/^(?:https?:)?\/\/[^/]+/, '');
  }

  /**
   * Return the current Droplet editor text content via __TestInterface.
   * Mirrors `the Droplet ACE text is "..."` from droplet_steps.rb.
   */
  async getDropletContents(): Promise<string> {
    return this.page.evaluate(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      () => (window as any).__TestInterface.getDropletContents() as string,
    );
  }

  /**
   * Return the current App Lab code via `Applab.getCode()`.
   * Mirrors `the droplet code is "..."` from applab.rb, which uses
   * `Applab.getCode()` — the canonical code getter regardless of block/text mode.
   * Unlike getDropletContents() or getAceEditorCode(), this includes the trailing
   * newline that App Lab appends to saved code.
   */
  async getCode(): Promise<string> {
    return this.page.evaluate(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      () => (window as any).Applab.getCode() as string,
    );
  }

  /**
   * Return the ACE editor text trimmed of leading/trailing whitespace.
   * Mirrors `ace editor code is equal to "..."` from droplet_steps.rb, which
   * calls `aceEditor.getValue().trim()` rather than `getDropletContents()`.
   */
  async getAceEditorCode(): Promise<string> {
    return this.page.evaluate(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const iface = (window as any).__TestInterface;
      return (iface.getDroplet().aceEditor.getValue() as string).trim();
    });
  }

  /**
   * Return App Lab's current level HTML (Applab.levelHtml).
   * Reflects the design-mode-authored HTML for the running app.
   */
  async getLevelHtml(): Promise<string> {
    return this.page.evaluate(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      () => (window as any).Applab.levelHtml as string,
    );
  }

  /**
   * Drag a design-mode element by type name into the app visualization.
   * Mirrors `I drag a {element_type} into the app` from applab.rb:
   * fires jQuery mousedown → mousemove → mouseup on the palette item.
   *
   * @param elementType - palette element type, e.g. 'BUTTON', 'SCREEN', 'LABEL', 'TEXT_AREA'
   */
  async dragElementToApp(elementType: string): Promise<void> {
    await this.page.evaluate((type: string) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const $ = (window as any).$;
      const el = $(`[data-element-type='${type}']`);
      const offset = el.offset();
      el.trigger(
        $.Event('mousedown', {which: 1, pageX: offset.left, pageY: offset.top}),
      );
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const vizOffset = ($('#visualization') as any).offset();
      $(document).trigger(
        $.Event('mousemove', {
          pageX: vizOffset.left + 15,
          pageY: vizOffset.top,
        }),
      );
      $(document).trigger(
        $.Event('mouseup', {
          pageX: vizOffset.left + 15,
          pageY: vizOffset.top,
        }),
      );
    }, elementType);
  }

  /**
   * Move a design-mode element to an exact app-space position.
   * Mirrors the App Lab property editor by calling Applab.updateProperty()
   * with the internal left/top property names.
   *
   * @param elementId - Design element id, e.g. `design_button1`
   * @param x - Left coordinate in app-space pixels
   * @param y - Top coordinate in app-space pixels
   */
  async setDesignElementPosition(
    elementId: string,
    x: number,
    y: number,
  ): Promise<void> {
    await this.page.evaluate(
      ({elementId, x, y}: {elementId: string; x: number; y: number}) => {
        const element = document.getElementById(elementId);
        if (!element) {
          throw new Error(
            `Could not find App Lab design element #${elementId}`,
          );
        }
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const applab = (window as any).Applab;
        applab.updateProperty(element, 'left', x);
        applab.updateProperty(element, 'top', y);
      },
      {elementId, x, y},
    );
  }

  /**
   * Save the current App Lab project through the same product API used by the
   * Cucumber `I save the project` step.
   */
  async saveProject(): Promise<void> {
    await this.waitForUiSaveAfter(() =>
      this.page.evaluate(() => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (window as any).Applab.serializeAndSave();
      }),
    );
  }

  /**
   * Dispatch a mousemove over an App Lab element so VisualizationOverlay
   * renders the user-visible coordinate/id tooltip.
   *
   * @param elementId - Element id to hover, without `#`
   */
  async hoverAppElement(elementId: string): Promise<void> {
    await this.page.evaluate((id: string) => {
      const element = document.getElementById(id);
      if (!element) {
        throw new Error(`Could not find App Lab element #${id}`);
      }
      const rect = element.getBoundingClientRect();
      const scale = rect.width / (element as HTMLElement).offsetWidth || 1;
      element.dispatchEvent(
        new MouseEvent('mousemove', {
          bubbles: true,
          cancelable: false,
          view: window,
          clientX: rect.left + 5 * scale,
          clientY: rect.top + 5 * scale,
        }),
      );
    }, elementId);
  }

  /**
   * Dispatch a mousemove over the App Lab visualization at app-space
   * coordinates, matching the Cucumber blank-screen hover step.
   *
   * @param x - App-space x coordinate
   * @param y - App-space y coordinate
   */
  async hoverVisualizationAt(x: number, y: number): Promise<void> {
    await this.page.evaluate(
      ({x, y}: {x: number; y: number}) => {
        const visualization = document.getElementById('visualizationOverlay');
        if (!visualization) {
          throw new Error('Could not find App Lab visualization overlay');
        }
        const rect = visualization.getBoundingClientRect();
        const scale =
          rect.width / (visualization as unknown as SVGElement).clientWidth ||
          1;
        visualization.dispatchEvent(
          new MouseEvent('mousemove', {
            bubbles: true,
            cancelable: false,
            view: window,
            clientX: rect.left + x * scale,
            clientY: rect.top + y * scale,
          }),
        );
      },
      {x, y},
    );
  }

  /**
   * Reset the project to its starting (initial) version via the Version History dialog.
   * Mirrors `I reset the puzzle to the starting version` from steps.rb.
   * Opens #versions-header, clicks "Start over", then confirms with #start-over-button.
   */
  async resetToStartingVersion(): Promise<void> {
    await this.page.locator('#versions-header').click();
    await this.page
      .locator('#showVersionsModal')
      .waitFor({state: 'visible', timeout: 15_000});
    // Scope to the modal — the main toolbar also has a "Start over" button
    // (#clear-puzzle-header) which would cause a strict-mode violation.
    const startOver = this.page
      .locator('#showVersionsModal')
      .locator('button', {hasText: 'Start over'});
    await startOver.waitFor({state: 'visible', timeout: 15_000});
    await startOver.click();
    await this.page
      .locator('#start-over-button')
      .waitFor({state: 'visible', timeout: 10_000});

    const navigationDone = this.page.waitForEvent('framenavigated', {
      predicate: frame => frame === this.page.mainFrame(),
      timeout: 60_000,
    });
    await this.page.locator('#start-over-button').click();
    await navigationDone;
    await this.page.waitForLoadState('domcontentloaded');
    await this.waitForReady();
    await expect(this.page.locator('.project_updated_at')).toContainText(
      'Saved',
      {timeout: 60_000},
    );
    await this.page
      .locator('#showVersionsModal')
      .waitFor({state: 'hidden', timeout: 15_000});
  }

  /**
   * Open the project share dialog, navigate to the Embed tab, and return the
   * embed page URL as a relative path (e.g. `/projects/applab/<id>/embed`).
   *
   * Mirrors the Cucumber `I navigate to the embedded version of my project`
   * flow from applab.rb: click .project_share → Show advanced options → Embed
   * → read the <iframe src="..."> value from the textarea.
   *
   * @param hideSource - when true, also ticks "Hide ability to view code"
   *   (mirrors `I navigate to the embedded version of my project with source hidden`)
   * @returns relative URL path suitable for page.goto()
   */
  async getEmbedUrl(hideSource = false): Promise<string> {
    await this.page.locator('.project_share').first().click();
    await this.page
      .locator('#project-share')
      .waitFor({state: 'visible', timeout: 15_000});

    // Expand advanced options to reveal the Embed tab.
    await this.page
      .locator('#project-share a', {hasText: 'Show advanced options'})
      .click();
    await this.page
      .locator('#project-share li', {hasText: 'Embed'})
      .waitFor({state: 'visible', timeout: 10_000});
    await this.page.locator('#project-share li', {hasText: 'Embed'}).click();

    if (hideSource) {
      // Tick the "Hide ability to view code" checkbox via its label.
      await this.page
        .locator('#project-share label', {hasText: 'Hide ability to view code'})
        .click();
    }

    // Read the <iframe src="..."> HTML from the embed textarea.
    const textarea = this.page.locator('#project-share textarea');
    await textarea.waitFor({state: 'visible', timeout: 10_000});
    const iframeHtml = await textarea.inputValue();

    // Close dialog.
    await this.page.keyboard.press('Escape');
    await this.page
      .locator('#project-share')
      .waitFor({state: 'hidden', timeout: 10_000});

    // Parse src attribute and strip origin so the path works against the test baseURL.
    const srcMatch = iframeHtml.match(/src="([^"]+)"/);
    if (!srcMatch) {
      throw new Error(`No src found in embed HTML: ${iframeHtml}`);
    }
    // Strip protocol + host (handles both "//host/path" and "https://host/path").
    return srcMatch[1].replace(/^(?:https?:)?\/\/[^/]+/, '');
  }
}
